mongodb/mongo_import.cc

```diff
@@ -0,0 +1,228 @@
+    addField("collection", &MongoImportConfig::collection,
+             "The collection to import");
+    addField("outputDataset", &MongoImportConfig::outputDataset,
+             "Output dataset configuration. This may refer either to an "
+             "existing dataset, or a fully specified but non-existing dataset "
+             "which will be created by the procedure.",
+             PolyConfigT<Dataset>().withType("sparse.mutable"));
+    addField("limit", &MongoImportConfig::limit,
+             "Maximum number of lines to process");
+    addField("offset", &MongoImportConfig::offset,
+             "Skip the first n lines.", int64_t(0));
+    addField("ignoreParsingErrors", &MongoImportConfig::ignoreParsingErrors,
+             "If true, any record causing an error will be skipped. Any "
+             "record with BSON regex or BSON internal data type will cause an "
+             "error.", false);
+    addField("select", &MongoImportConfig::select,
+             "Which columns to use.",
+             SelectExpression::STAR);
+    addField("where", &MongoImportConfig::where,
+             "Which lines to use to create rows.",
+             SqlExpression::TRUE);
+    addField("named", &MongoImportConfig::named,
+             "Row name expression for output dataset. Note that each row "
+             "must have a unique name and that names cannot be objects. The "
+             "default value, `oid()`, refers to the MongoDB ObjectID.",
+             SqlExpression::parse("oid()"));
+
+    onPostValidate = [] (MongoImportConfig * config,
+                         JsonParsingContext & context)
+    {
+        validateConnectionScheme(config->connectionScheme);
+        validateCollection(config->collection);
+    };
+}
+
+struct MongoImportProcedure: public Procedure {
+    MongoImportProcedure(MldbServer * server,
+                         PolyConfig config_,
+                         std::function<bool (Json::Value)> onProgress)
+        : Procedure(server)
+    {
+        config = config_.params.convert<MongoImportConfig>();
+    }
+
+    Any getStatus() const override
+    {
+        Json::Value result;
+        result["ok"] = true;
+        return result;
+    }
+    MongoImportConfig config;
+
+    RunOutput run(const ProcedureRunConfig & run,
+                  const std::function<bool (const Json::Value &)> & onProgress) const override
+    {
+        const auto runConfig = applyRunConfOverProcConf(config, run);
+
+        mongocxx::uri mongoUri(runConfig.connectionScheme);
+        mongocxx::client conn(mongoUri);
+        auto db = conn[mongoUri.database()];
+
+        DEBUG_MSG(logger) << "\n"
+            << "Db name:    " << mongoUri.database() << "\n"
+            << "Collection: " << runConfig.collection;
+
+        auto output = createDataset(server, runConfig.outputDataset,
+                                    nullptr, true /*overwrite*/);
+
+        MongoScope mongoScope(server);
+        ExpressionValue storage;
```

Comment:
This is shared across several threads.  I believe you fixed something similar last week that was causing a crash.